x-pack/filebeat/input/httpjson/request.go

```diff
@@ -146,38 +167,177 @@ func (rf *requestFactory) newHTTPRequest(stdCtx context.Context, trCtx *transfor
+		if i == 0 {
+			// perform and store regular call responses
+			httpResp, err = rf.collectResponse(stdCtx, trCtx, r, publisher)
+			if err != nil {
+				return fmt.Errorf("failed to execute rf.collectResponse: %w", err)
+			}
+			if rfSize == 1 {
+				finalResps = append(finalResps, httpResp)
+				n, err = r.processAndPublishEvents(stdCtx, trCtx, publisher, finalResps, true)
+				if err != nil {
+					return err
+				}
+				continue
+			} else {
+				intermediateResps = append(intermediateResps, httpResp)
+				ids, err = r.getIdsFromResponses(intermediateResps, r.requestFactories[i+1].replace)
+				if err != nil {
+					return err
+				}
+				n, err = r.processAndPublishEvents(stdCtx, trCtx, publisher, intermediateResps, false)
+				if err != nil {
+					return err
+				}
+			}
+		} else {
+			if len(ids) == 0 {
+				n = 0
+				continue
+			}
+			urlCopy = rf.url
+			urlString = rf.url.String()
+			// perform request over collected ids
+			for _, id := range ids {
+				// reformat urls of requestFactory using ids
+				rf.url, err = generateNewUrl(rf.replace, urlString, id)
+				if err != nil {
+					return fmt.Errorf("failed to generate new URL: %w", err)
+				}
+
+				// collect data from new urls
+				httpResp, err = rf.collectResponse(stdCtx, trCtx, r, publisher)
+				if err != nil {
+					return fmt.Errorf("failed to execute rf.collectResponse: %w", err)
+				}
+				// store data according to response type
+				if i == rfSize-1 && len(ids) != 0 {
+					finalResps = append(finalResps, httpResp)
+				} else {
+					intermediateResps = append(intermediateResps, httpResp)
+				}
+			}
+			rf.url = urlCopy
+
+			if i == rfSize-1 {
+				split = r.responseProcessor.split
+				r.responseProcessor.split = rf.split
+				n, err = r.processAndPublishEvents(stdCtx, trCtx, publisher, finalResps, true)
+				if err != nil {
+					return err
+				}
+				r.responseProcessor.split = split
+				continue
+			} else {
+				ids, err = r.getIdsFromResponses(intermediateResps, r.requestFactories[i+1].replace)
+				if err != nil {
+					return err
+				}
+				split = r.responseProcessor.split
+				r.responseProcessor.split = rf.split
+				n, err = r.processAndPublishEvents(stdCtx, trCtx, publisher, intermediateResps, false)
+				if err != nil {
+					return err
+				}
+				r.responseProcessor.split = split
+			}
+		}
 	}
+
 	defer httpResp.Body.Close()
+	r.log.Infof("request finished: %d events published", n)
 
-	eventsCh, err := r.responseProcessor.startProcessing(stdCtx, trCtx, httpResp)
+	return nil
+}
+
+// getIdsFromResponses returns ids from responses
+func (r *requester) getIdsFromResponses(intermediateResps []*http.Response, replace string) ([]string, error) {
+	var b []byte
+	var ids []string
+	var err error
+	// collect ids from all responses
+	for _, resp := range intermediateResps {
+		if resp.Body != nil {
+			b, err = io.ReadAll(resp.Body)
+			if err != nil {
+				return nil, fmt.Errorf("error while reading response body: %w", err)
+			}
+		}
```

Comment:
> But that makes me wonder if you really need access to the http.Response here. Maybe we could update requestFactory.collectResponse to return a []byte instead of the http.Response. That way the only bit of code that needs to worry about the quirks of HTTP responses (like closing the response body) is requestFactory.collectResponse. What do you think?
Right, but I am thinking like will it not cause problems for r.responseProcessor.startProcessing? it is using HTTP responses to process in pagination and all. so before closing it may cause problems for responseProcessor.

Reply:
> Another thing I noticed is that there are no checks for the HTTP status code (or I missed them). The HTTP client we use does return an error if the status code is >= 400, that means response in the 1xx and 3xx ranges could reach this code path and will error because the data is not in the format we want or there is no data.
Is that by design? Or should we do some validation of the status code?

I think it is by design as I have noticed while collecting responses it is storing all types of responses even if it's not required.